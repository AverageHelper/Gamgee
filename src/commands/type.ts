import type { Command } from "./Command";
import { deleteMessage } from "../actions/messages/deleteMessage";

const type: Command = {
  name: "t",
  description: "Start typing. :wink:",
  async execute({ message, client, logger }) {
    const channel = message.channel;

    logger.debug(
      `${client.user?.username.concat(" is") ?? "I am"} typing in channel ${channel.id}...`
    );
    await deleteMessage(message, "Spam: Users don't need to see this command run");

    setTimeout(() => {
      channel.stopTyping(true);
    }, 5000);

    await channel.startTyping();
    logger.debug(`Finished typing in channel ${channel.id}`);
  }
};

export default type;
