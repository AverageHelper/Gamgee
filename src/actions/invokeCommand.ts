import type Discord from "discord.js";
import type { Command, CommandContext, GuildedCommand, Subcommand } from "../commands/index.js";
import { Collection, Permissions } from "discord.js";
import { isGuildedCommandContext } from "../commands/index.js";
import { useLogger } from "../logger.js";
import { userHasPermissionInChannel, userHasRoleInGuild } from "../userHasOneOfRoles.js";
import { richErrorMessage } from "../helpers/richErrorMessage.js";

const logger = useLogger();

type Invocable = Command | Subcommand;

async function failPermissions(context: CommandContext): Promise<void> {
	return context.replyPrivately("You don't have permission to run that command.");
}

async function failNoGuild(context: CommandContext): Promise<void> {
	return context.reply({ content: "Can't do that here.", ephemeral: true });
}

function neverFallthrough(val: never): never {
	throw new TypeError(`Unexpected case: ${JSON.stringify(val)}`);
}

/**
 * Assesses whether the calling guild member is allowed to run the given command.
 *
 * @param member The calling guild member.
 * @param command The command the user wants to run.
 * @param channel The channel in which the command is being run.
 *
 * @returns `true` if the user may be permitted to run the command. `false` otherwise.
 */
export async function assertUserCanRunCommand(
	member: Discord.GuildMember,
	command: GuildedCommand,
	channel: Discord.GuildTextBasedChannel
): Promise<boolean> {
	// Gather permission details
	const guild = channel.guild;
	const guildId = guild.id;
	const client = guild.client;

	// TODO: Cache this for each command on startup
	const guildCommands = await guild.commands
		// eslint-disable-next-line @typescript-eslint/consistent-type-assertions
		.fetch({ command: client.user?.id } as Discord.FetchGuildApplicationCommandFetchOptions) // FIXME: wating on v14 for this call to work
		.catch((error: unknown) => {
			logger.error(richErrorMessage(`Failed to fetch commands for guild ${guildId}`, error));
			return new Collection<string, Discord.ApplicationCommand>();
		});
	const guildCommand = guildCommands.find(cmd => cmd.name === command.name) ?? null;

	const defaultPermissions =
		command.defaultMemberPermissions !== undefined
			? new Permissions(command.defaultMemberPermissions)
			: null;
	// TODO: Cache this for each command on startup
	const permissions = await guildCommand?.permissions //
		.fetch({})
		.catch((error: unknown) => {
			logger.error(
				richErrorMessage(
					`Failed to fetch permissions for command '${guildCommand.name}' in guild ${guildId}`,
					error
				)
			);
			return undefined;
		});

	if (!permissions && !defaultPermissions) {
		// No permissions configured, assume the user is fine
		logger.debug(`Command '${command.name}' has no permission requirements to satisfy.`);
		logger.debug("\tProceeding...");
		return true;
	}

	if (permissions?.length === 0 || defaultPermissions?.toArray().length === 0) {
		// Empty permissions configured, assume no access
		logger.debug(`Command '${command.name}' is configured to block any access.`);
		return false;
	}

	logger.debug(
		`Command '${command.name}' requires that callers satisfy 1 of ${
			permissions?.length ?? "null"
		} cases:`
	);

	// Check configured permissions
	let idx = 0;
	for (const access of permissions ?? []) {
		idx += 1;
		logger.debug(
			`\tCase ${idx}: User must${access.permission ? "" : " not"} have ${access.type} ID: ${
				access.id
			}`
		);
		switch (access.type) {
			case "ROLE": {
				const userHasRole = await userHasRoleInGuild(member, access.id, guild);
				logger.debug(`\tUser ${userHasRole ? "has" : "does not have"} role ${access.id}`);
				if (access.permission && userHasRole) {
					logger.debug("\tProceeding...");
					return true;
				}
				break;
			}

			case "USER": {
				const userHasId = member.user.id === access.id;
				logger.debug(`\tUser ${userHasId ? "has" : "does not have"} ID ${access.id}`);
				if (access.permission && userHasId) {
					logger.debug("\tProceeding...");
					return true;
				}
				break;
			}

			default:
				return neverFallthrough(access.type);
		}
	}

	// Check default permissions
	if (defaultPermissions && userHasPermissionInChannel(member, defaultPermissions, channel)) {
		return true;
	}

	// Caller fails permissions checks
	logger.debug("User did not pass any permission checks.");
	return false;
}

/**
 * Runs the command if the invocation context satisfies the command's
 * declared guild and permission requirements.
 *
 * @param command The command to execute.
 * @param context The invocation context.
 */
export async function invokeCommand(command: Invocable, context: CommandContext): Promise<void> {
	if (!command.requiresGuild) {
		// No guild required
		logger.debug(`Command '${command.name}' does not require guild information.`);
		logger.debug("Proceeding...");
		return command.execute(context);
	}

	if (!isGuildedCommandContext(context)) {
		// No guild found
		logger.debug(`Command '${command.name}' requires guild information, but none was found.`);
		return failNoGuild(context);
	}

	if (
		context.channel &&
		command.type !== "SUB_COMMAND" &&
		(await assertUserCanRunCommand(context.member, command, context.channel))
	) {
		return command.execute(context);
	}
	return failPermissions(context);
}
