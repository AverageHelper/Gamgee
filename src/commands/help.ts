import type { Command } from "./Command.js";
import { composed, createPartialString, push, pushNewLine } from "../helpers/composeStrings.js";
import { DEFAULT_LOCALE, isSupportedLocale, localizations, t } from "../i18n.js";
import { describeAllCommands } from "../actions/describeAllCommands.js";

export const help: Command = {
	name: "help",
	nameLocalizations: localizations("commands.help.name"),
	description: "Print a handy help message.",
	descriptionLocalizations: localizations("commands.help.description"),
	requiresGuild: false,
	async execute(context) {
		// Dynamic import here, b/c ./index depends on us to resolve
		const { allCommands } = await import("./index.js");

		let locale = context.userLocale ?? DEFAULT_LOCALE;
		if (!isSupportedLocale(locale)) {
			locale = DEFAULT_LOCALE;
		}

		const descriptions = await describeAllCommands(context, allCommands, locale);
		const response = createPartialString(t("commands.help.response", locale));
		pushNewLine(response);
		push(descriptions, response);
		return await context.replyPrivately(composed(response));
	}
};
