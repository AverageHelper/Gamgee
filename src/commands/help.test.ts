jest.mock("../actions/invokeCommand");

import { assertUserCanRunCommand } from "../actions/invokeCommand.js";
import { Permissions } from "discord.js";
const mockAssertUserCanRunCommand = assertUserCanRunCommand as jest.Mock;

import type { GuildedCommand, GuildedCommandContext } from "./Command.js";
import { help } from "./help.js";

const mockReplyPrivately = jest.fn();

describe("Help command", () => {
	let context: GuildedCommandContext;

	beforeEach(() => {
		context = {
			type: "message",
			guild: {
				id: "the-guild"
			},
			storage: null,
			replyPrivately: mockReplyPrivately
		} as unknown as GuildedCommandContext;

		mockAssertUserCanRunCommand.mockResolvedValue(true);
	});

	test("describes all commands", async () => {
		await help.execute(context);
		expect(mockReplyPrivately).toHaveBeenCalledTimes(1);
		expect(mockReplyPrivately).toHaveBeenCalledWith(expect.toBeString());

		const calls = mockReplyPrivately.mock.calls[0] as Array<unknown>;
		const description = calls[0];
		expect(description).toMatchSnapshot();
	});

	test("describe pleb commands", async () => {
		mockAssertUserCanRunCommand.mockImplementation((user, command: GuildedCommand) => {
			if (new Permissions(command.defaultMemberPermissions).has("ADMINISTRATOR")) {
				return Promise.resolve(false);
			}
			return Promise.resolve(true);
		});

		await help.execute(context);
		expect(mockReplyPrivately).toHaveBeenCalledTimes(1);
		expect(mockReplyPrivately).toHaveBeenCalledWith(expect.toBeString());

		const calls = mockReplyPrivately.mock.calls[0] as Array<unknown>;
		const description = calls[0];
		expect(description).toMatchSnapshot();
	});
});
