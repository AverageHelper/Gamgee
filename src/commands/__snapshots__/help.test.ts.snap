// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Help command describe pleb commands 1`] = `
"Commands:
\`?cooldown\` - Find out when you can submit again.
\`?help\` - Print a handy help message.
\`?howto\` - Print instructions for using the common queue commands.
\`?languages\` - Print my core repository's language statistics.
\`?limits\` - Display the song queue's submission limits.
\`?now-playing\` OR \`?nowplaying\` - Reveal the current song in the queue (or my best guess).
\`?ping\` - Ping my host server to check latency.
\`?sr [url]\` - Submit a song to the queue.
\`?stats\` - Get your personal queue statistics.
\`?test\` - Make sure I still know how to talk to video services.
\`?t\` - Start a typing indicator.
\`?userinfo\` - Returns some information about you.
\`?version\` - Display the bot's current codebase version.
\`?video [url]\` - Reply with the video title and duration.
"
`;

exports[`Help command describes all commands 1`] = `
"Commands:
\`?config\` - Read and modify config options.
    \`?config get <command_prefix>\` - Get the value of a configuration setting.
    \`?config set <command_prefix> [value]\` - Set the value of a configuration setting.
    \`?config unset <command_prefix>\` - Reset the value of a configuration setting to default.
\`?cooldown\` - Find out when you can submit again.
\`?help\` - Print a handy help message.
\`?howto\` - Print instructions for using the common queue commands.
\`?languages\` - Print my core repository's language statistics.
\`?limits\` - Display the song queue's submission limits.
\`?now-playing\` OR \`?nowplaying\` - Reveal the current song in the queue (or my best guess).
\`?ping\` - Ping my host server to check latency.
\`?quo\` - Administrative commands to manage the song queue.
    \`?quo setup [channel]\` - Set a channel as the 'queue' channel.
    \`?quo teardown\` - Deletes and un-sets the current queue.
    \`?quo blacklist [?user]\` - Show the list of blacklisted users, or add a user to the blacklist.
    \`?quo whitelist [user]\` - Allows a previously-blacklisted user to make song requests.
    \`?quo open\` - Start accepting song requests to the queue.
    \`?quo close\` - Stop accepting song requests to the queue.
    \`?quo limit [?count | cooldown | entry-duration-min | entry-duration | queue-duration] [?value]\` - Set a limit value on the queue. (Time in seconds, where applicable)
    \`?quo stats\` - Print statistics on the current queue.
    \`?quo restart\` - Empty the queue and start a fresh queue session.
\`?sr [url]\` - Submit a song to the queue.
\`?stats\` - Get your personal queue statistics.
\`?test\` - Make sure I still know how to talk to video services.
\`?t\` - Start a typing indicator.
\`?userinfo\` - Returns some information about you.
\`?version\` - Display the bot's current codebase version.
\`?video [url]\` - Reply with the video title and duration.
"
`;

exports[`Help command describes all commands in de 1`] = `
"Befehle:
\`?config\` - Read and modify config options.
    \`?config get <command_prefix>\` - Get the value of a configuration setting.
    \`?config set <command_prefix> [value]\` - Set the value of a configuration setting.
    \`?config unset <command_prefix>\` - Reset the value of a configuration setting to default.
\`?cooldown\` - Finden Sie heraus, wann Sie wieder einreichen können.
\`?hilfe\` - Drucken Sie eine nützliche Hilfenachricht.
\`?wieman\` - Print instructions for using the common queue commands.
\`?languages\` - Print my core repository's language statistics.
\`?limits\` - Display the song queue's submission limits.
\`?now-playing\` OR \`?nowplaying\` - Reveal the current song in the queue (or my best guess).
\`?ping\` - Ping my host server to check latency.
\`?quo\` - Administrative commands to manage the song queue.
    \`?quo setup [channel]\` - Set a channel as the 'queue' channel.
    \`?quo teardown\` - Deletes and un-sets the current queue.
    \`?quo blacklist [?user]\` - Show the list of blacklisted users, or add a user to the blacklist.
    \`?quo whitelist [user]\` - Allows a previously-blacklisted user to make song requests.
    \`?quo open\` - Start accepting song requests to the queue.
    \`?quo close\` - Stop accepting song requests to the queue.
    \`?quo limit [?count | cooldown | entry-duration-min | entry-duration | queue-duration] [?value]\` - Set a limit value on the queue. (Time in seconds, where applicable)
    \`?quo stats\` - Print statistics on the current queue.
    \`?quo restart\` - Empty the queue and start a fresh queue session.
\`?sr [url]\` - Senden Sie einen Song an die Warteschlange.
\`?stats\` - Get your personal queue statistics.
\`?test\` - Make sure I still know how to talk to video services.
\`?t\` - Start a typing indicator.
\`?benutzerinfo\` - Gibt einige Informationen über Sie zurück.
\`?version\` - Display the bot's current codebase version.
\`?video [url]\` - Reply with the video title and duration.
"
`;

exports[`Help command describes all commands in en-GB 1`] = `
"Commands:
\`?config\` - Read and modify config options.
    \`?config get <command_prefix>\` - Get the value of a configuration setting.
    \`?config set <command_prefix> [value]\` - Set the value of a configuration setting.
    \`?config unset <command_prefix>\` - Reset the value of a configuration setting to default.
\`?cooldown\` - Find out when you can submit again.
\`?help\` - Print a useful help message.
\`?howto\` - Print instructions for using the common queue commands.
\`?languages\` - Print my core repository's language statistics.
\`?limits\` - Display the song queue's submission limits.
\`?now-playing\` OR \`?nowplaying\` - Reveal the current song in the queue (or my best guess).
\`?ping\` - Ping my host server to check latency.
\`?quo\` - Administrative commands to manage the song queue.
    \`?quo setup [channel]\` - Set a channel as the 'queue' channel.
    \`?quo teardown\` - Deletes and un-sets the current queue.
    \`?quo blacklist [?user]\` - Show the list of blacklisted users, or add a user to the blacklist.
    \`?quo whitelist [user]\` - Allows a previously-blacklisted user to make song requests.
    \`?quo open\` - Start accepting song requests to the queue.
    \`?quo close\` - Stop accepting song requests to the queue.
    \`?quo limit [?count | cooldown | entry-duration-min | entry-duration | queue-duration] [?value]\` - Set a limit value on the queue. (Time in seconds, where applicable)
    \`?quo stats\` - Print statistics on the current queue.
    \`?quo restart\` - Empty the queue and start a fresh queue session.
\`?sr [url]\` - Submit a song to the queue.
\`?stats\` - Get your personal queue statistics.
\`?test\` - Make sure I still know how to talk to video services.
\`?t\` - Start a typing indicator.
\`?userinfo\` - Returns some information about you.
\`?version\` - Display the bot's current codebase version.
\`?video [url]\` - Reply with the video title and duration.
"
`;

exports[`Help command describes all commands in en-US 1`] = `
"Commands:
\`?config\` - Read and modify config options.
    \`?config get <command_prefix>\` - Get the value of a configuration setting.
    \`?config set <command_prefix> [value]\` - Set the value of a configuration setting.
    \`?config unset <command_prefix>\` - Reset the value of a configuration setting to default.
\`?cooldown\` - Find out when you can submit again.
\`?help\` - Print a handy help message.
\`?howto\` - Print instructions for using the common queue commands.
\`?languages\` - Print my core repository's language statistics.
\`?limits\` - Display the song queue's submission limits.
\`?now-playing\` OR \`?nowplaying\` - Reveal the current song in the queue (or my best guess).
\`?ping\` - Ping my host server to check latency.
\`?quo\` - Administrative commands to manage the song queue.
    \`?quo setup [channel]\` - Set a channel as the 'queue' channel.
    \`?quo teardown\` - Deletes and un-sets the current queue.
    \`?quo blacklist [?user]\` - Show the list of blacklisted users, or add a user to the blacklist.
    \`?quo whitelist [user]\` - Allows a previously-blacklisted user to make song requests.
    \`?quo open\` - Start accepting song requests to the queue.
    \`?quo close\` - Stop accepting song requests to the queue.
    \`?quo limit [?count | cooldown | entry-duration-min | entry-duration | queue-duration] [?value]\` - Set a limit value on the queue. (Time in seconds, where applicable)
    \`?quo stats\` - Print statistics on the current queue.
    \`?quo restart\` - Empty the queue and start a fresh queue session.
\`?sr [url]\` - Submit a song to the queue.
\`?stats\` - Get your personal queue statistics.
\`?test\` - Make sure I still know how to talk to video services.
\`?t\` - Start a typing indicator.
\`?userinfo\` - Returns some information about you.
\`?version\` - Display the bot's current codebase version.
\`?video [url]\` - Reply with the video title and duration.
"
`;

exports[`Help command describes all commands in es-ES 1`] = `
"Comandos:
\`?config\` - Read and modify config options.
    \`?config get <command_prefix>\` - Get the value of a configuration setting.
    \`?config set <command_prefix> [value]\` - Set the value of a configuration setting.
    \`?config unset <command_prefix>\` - Reset the value of a configuration setting to default.
\`?cooldown\` - Averigüe cuándo puede enviar de nuevo.
\`?ayuda\` - Imprime un mensaje de ayuda útil.
\`?cómo\` - Print instructions for using the common queue commands.
\`?languages\` - Print my core repository's language statistics.
\`?limits\` - Display the song queue's submission limits.
\`?now-playing\` OR \`?nowplaying\` - Reveal the current song in the queue (or my best guess).
\`?ping\` - Ping my host server to check latency.
\`?quo\` - Administrative commands to manage the song queue.
    \`?quo setup [channel]\` - Set a channel as the 'queue' channel.
    \`?quo teardown\` - Deletes and un-sets the current queue.
    \`?quo blacklist [?user]\` - Show the list of blacklisted users, or add a user to the blacklist.
    \`?quo whitelist [user]\` - Allows a previously-blacklisted user to make song requests.
    \`?quo open\` - Start accepting song requests to the queue.
    \`?quo close\` - Stop accepting song requests to the queue.
    \`?quo limit [?count | cooldown | entry-duration-min | entry-duration | queue-duration] [?value]\` - Set a limit value on the queue. (Time in seconds, where applicable)
    \`?quo stats\` - Print statistics on the current queue.
    \`?quo restart\` - Empty the queue and start a fresh queue session.
\`?sr [url]\` - Envía una canción a la cola.
\`?stats\` - Get your personal queue statistics.
\`?test\` - Make sure I still know how to talk to video services.
\`?t\` - Start a typing indicator.
\`?usuarioinfo\` - Devuelve alguna información sobre usted.
\`?version\` - Display the bot's current codebase version.
\`?video [url]\` - Reply with the video title and duration.
"
`;

exports[`Help command describes all commands in fr 1`] = `
"Commandes:
\`?config\` - Read and modify config options.
    \`?config get <command_prefix>\` - Get the value of a configuration setting.
    \`?config set <command_prefix> [value]\` - Set the value of a configuration setting.
    \`?config unset <command_prefix>\` - Reset the value of a configuration setting to default.
\`?cooldown\` - Découvrez quand vous pourrez soumettre à nouveau.
\`?aider\` - Imprimez un message d'aide utile.
\`?comment\` - Print instructions for using the common queue commands.
\`?languages\` - Print my core repository's language statistics.
\`?limits\` - Display the song queue's submission limits.
\`?now-playing\` OR \`?nowplaying\` - Reveal the current song in the queue (or my best guess).
\`?ping\` - Ping my host server to check latency.
\`?quo\` - Administrative commands to manage the song queue.
    \`?quo setup [channel]\` - Set a channel as the 'queue' channel.
    \`?quo teardown\` - Deletes and un-sets the current queue.
    \`?quo blacklist [?user]\` - Show the list of blacklisted users, or add a user to the blacklist.
    \`?quo whitelist [user]\` - Allows a previously-blacklisted user to make song requests.
    \`?quo open\` - Start accepting song requests to the queue.
    \`?quo close\` - Stop accepting song requests to the queue.
    \`?quo limit [?count | cooldown | entry-duration-min | entry-duration | queue-duration] [?value]\` - Set a limit value on the queue. (Time in seconds, where applicable)
    \`?quo stats\` - Print statistics on the current queue.
    \`?quo restart\` - Empty the queue and start a fresh queue session.
\`?sr [url]\` - Soumettre une chanson à la file d'attente.
\`?stats\` - Get your personal queue statistics.
\`?test\` - Make sure I still know how to talk to video services.
\`?t\` - Start a typing indicator.
\`?utilinfo\` - Retourne quelques informations sur vous.
\`?version\` - Display the bot's current codebase version.
\`?video [url]\` - Reply with the video title and duration.
"
`;

exports[`Help command describes all commands in hu 1`] = `
"Parancsok:
\`?config\` - Read and modify config options.
    \`?config get <command_prefix>\` - Get the value of a configuration setting.
    \`?config set <command_prefix> [value]\` - Set the value of a configuration setting.
    \`?config unset <command_prefix>\` - Reset the value of a configuration setting to default.
\`?cooldown\` - Tudja meg, mikor küldheti be újra.
\`?segítség\` - Nyomtasson ki egy hasznos súgóüzenetet.
\`?hogyankell\` - Print instructions for using the common queue commands.
\`?languages\` - Print my core repository's language statistics.
\`?limits\` - Display the song queue's submission limits.
\`?now-playing\` OR \`?nowplaying\` - Reveal the current song in the queue (or my best guess).
\`?ping\` - Ping my host server to check latency.
\`?quo\` - Administrative commands to manage the song queue.
    \`?quo setup [channel]\` - Set a channel as the 'queue' channel.
    \`?quo teardown\` - Deletes and un-sets the current queue.
    \`?quo blacklist [?user]\` - Show the list of blacklisted users, or add a user to the blacklist.
    \`?quo whitelist [user]\` - Allows a previously-blacklisted user to make song requests.
    \`?quo open\` - Start accepting song requests to the queue.
    \`?quo close\` - Stop accepting song requests to the queue.
    \`?quo limit [?count | cooldown | entry-duration-min | entry-duration | queue-duration] [?value]\` - Set a limit value on the queue. (Time in seconds, where applicable)
    \`?quo stats\` - Print statistics on the current queue.
    \`?quo restart\` - Empty the queue and start a fresh queue session.
\`?sr [url]\` - Adjon fel egy dalt a sorba.
\`?stats\` - Get your personal queue statistics.
\`?test\` - Make sure I still know how to talk to video services.
\`?t\` - Start a typing indicator.
\`?felhasználóiadatok\` - Néhány információt visszaad rólad.
\`?version\` - Display the bot's current codebase version.
\`?video [url]\` - Reply with the video title and duration.
"
`;

exports[`Help command describes all commands in pt-BR 1`] = `
"Comandos:
\`?config\` - Read and modify config options.
    \`?config get <command_prefix>\` - Get the value of a configuration setting.
    \`?config set <command_prefix> [value]\` - Set the value of a configuration setting.
    \`?config unset <command_prefix>\` - Reset the value of a configuration setting to default.
\`?cooldown\` - Descubra quando você pode enviar novamente.
\`?ajuda\` - Imprima uma mensagem de ajuda útil.
\`?como\` - Print instructions for using the common queue commands.
\`?languages\` - Print my core repository's language statistics.
\`?limits\` - Display the song queue's submission limits.
\`?now-playing\` OR \`?nowplaying\` - Reveal the current song in the queue (or my best guess).
\`?ping\` - Ping my host server to check latency.
\`?quo\` - Administrative commands to manage the song queue.
    \`?quo setup [channel]\` - Set a channel as the 'queue' channel.
    \`?quo teardown\` - Deletes and un-sets the current queue.
    \`?quo blacklist [?user]\` - Show the list of blacklisted users, or add a user to the blacklist.
    \`?quo whitelist [user]\` - Allows a previously-blacklisted user to make song requests.
    \`?quo open\` - Start accepting song requests to the queue.
    \`?quo close\` - Stop accepting song requests to the queue.
    \`?quo limit [?count | cooldown | entry-duration-min | entry-duration | queue-duration] [?value]\` - Set a limit value on the queue. (Time in seconds, where applicable)
    \`?quo stats\` - Print statistics on the current queue.
    \`?quo restart\` - Empty the queue and start a fresh queue session.
\`?sr [url]\` - Envie uma música para a fila.
\`?stats\` - Get your personal queue statistics.
\`?test\` - Make sure I still know how to talk to video services.
\`?t\` - Start a typing indicator.
\`?usuárioinfo\` - Retorna algumas informações sobre você.
\`?version\` - Display the bot's current codebase version.
\`?video [url]\` - Reply with the video title and duration.
"
`;
